src/duckling/canvas/tools/entity-move-tool.ts

```diff
@@ -23,83 +22,316 @@ import {BaseTool, CanvasMouseEvent, CanvasKeyEvent} from './base-tool';
-        }
-
-        let drawnConstruct = new EntityMoveToolDrawnConstruct(canvasZoom, box);
+        let drawnConstruct = new EntityMoveToolDrawnConstruct(
+            canvasZoom, 
+            this._drawSelectionBox(), 
+            this._drawSelectedEntityGizmos());
         drawnConstruct.layer = Number.POSITIVE_INFINITY;
         return drawnConstruct;
+    }
 
+    private _drawSelectedEntityGizmos() : Box2[] {
+        let boxes : Box2[] = [];
+        for (let selection of this._selectionService.selections.value) {
+            let entityBox = this._entityBoxService.getEntityBox(selection.key);
+            if (entityBox) {
+                boxes.push(entityBox);
+            }
+        }
+        let allEntitiesBox = this._entityBoxService.getEntitiesBoundingBox(
+            this._selectionService.selections.value.map((selection: Selection) => { return selection.key; }));
+        if (allEntitiesBox) {
+            boxes.push(allEntitiesBox);
+        }
+        return boxes;
+    }
+
+    private _drawSelectionBox() : Box2 {
+        if (!this._initialMouseLocation || !this._selectionBoxDimensions) {
+            return;
+        }
+
+        if (this._state !== State.clickSelectionBox && this._state !== State.shiftClickSelectionBox) {
+            return;
+        }
+
+        return {
+            position: this._initialMouseLocation,
+            dimension: this._selectionBoxDimensions,
+            rotation: 0
+        };
     }
 
     onStageDown(event : CanvasMouseEvent) {
-        this._selectedEntityKey = this._entitySelectionService.getEntityKey(event.stageCoords);
+        console.log("Initial: " + State[this._state]);
         this._mergeKey = newMergeKey();
-        this._selectionService.select(this._selectedEntityKey, this._mergeKey);
+        this._initialMouseLocation = event.stageCoords;
+        
+        let selectedEntityKey = this._selectionService.getEntityKeyAtPosition(event.stageCoords);
+
+        this._setStateFromMouseDown(selectedEntityKey, event.shiftKey)
+
+        console.log("Down: " + State[this._state]);
+
+        switch (this._state) {
+            case State.clickOnUnselected : {
+                this._selectionService.deselect(this._mergeKey);
+                this._selectionService.select([selectedEntityKey], this._mergeKey);
+                break;                
+            }
+            case State.shiftClickOnUnselected: {
+                this._selectionService.select([selectedEntityKey], this._mergeKey);
+                break;
+            }
+        }
+        this._cacheEntityKeyAtMouseDownPosition(selectedEntityKey);
+        this._cacheInitialPositions();
+    }
+
+    private _cacheEntityKeyAtMouseDownPosition(selectedEntityKey: EntityKey) : void {
+        switch (this._state) {
+            case State.clickOnSelected:
+            case State.clickOnUnselected:
+            case State.shiftClickOnSelected:
+            case State.shiftClickOnUnselected: {
```

Comment:
Hmmm. So in order to be consistent, I would have to remove all of the block scoping for all of the switch cases throughout the file. In onStageUp we do actually declare variables within the block scoped cases. So I'd prefer to just keep the blocks. If you feel strongly, I can change it and declare the variable at the top of the function. It's not too bad; I'd just rather not